paddle/fluid/framework/details/nccl_op_handle.h

```diff
@@ -0,0 +1,230 @@
+  }
+
+  void NCCLAllReduce(platform::Place place, const void* sendbuff,
+                     void* recvbuff, size_t count, ncclDataType_t datatype,
+                     ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    if (!use_hierarchical_allreduce_) {
+      FlatNcclAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+      return;
+    }
+
+    HierarchicalAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+  }
+
+  void HierarchicalAllReduce(platform::Place place, const void* sendbuff,
+                             void* recvbuff, size_t count,
+                             ncclDataType_t datatype, ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    InterAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+    if (nccl_ctxs_->NeedExterAllReduce()) {
+      ExterAllReduce(place, recvbuff, recvbuff, count, datatype, op);
+    }
+    InterBroadCast(place, recvbuff, count, datatype, op);
+  }
+
+ protected:
+  void InterAllReduce(platform::Place place, const void* sendbuff,
+                      void* recvbuff, size_t count, ncclDataType_t datatype,
+                      ncclRedOp_t op) {
+    auto nccl_ctxs = nccl_ctxs_->GetHierarchicalInterCtx(run_order_);
+    int dev_id = boost::get<platform::CUDAPlace>(place).device;
+    auto& nccl_ctx = nccl_ctxs->at(dev_id);
+    auto stream = nccl_ctx.stream();
+    auto comm = nccl_ctx.comm_;
+
+    VLOG(10) << "before all reduce"
+             << " run_order:" << run_order_ << ", buffer:" << sendbuff
+             << ", numel:" << count << ", dev_id:" << dev_id
+             << ", dtype:" << datatype << ", place:" << place
+             << ", stream:" << stream;
+
+    PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
+        sendbuff, recvbuff, count, datatype, ncclSum, comm, stream));
+    if (FLAGS_sync_nccl_allreduce) {
+      PADDLE_ENFORCE(cudaStreamSynchronize(stream),
+                     "sync HierarchicalAllReduce inter stream error");
+    } else {
+      cudaEventRecord(inter_events_.at(dev_id), stream);
+    }
+  }
+
+  void ExterAllReduce(platform::Place place, const void* sendbuff,
+                      void* recvbuff, size_t count, ncclDataType_t datatype,
+                      ncclRedOp_t op) {
+    auto nccl_ctxs = nccl_ctxs_->GetHierarchicalExterCtx(run_order_);
+    PADDLE_ENFORCE(nccl_ctxs_, "can't get exter %d nccl_ctxs", run_order_);
+    int dev_id = boost::get<platform::CUDAPlace>(place).device;
+    auto& nccl_ctx = nccl_ctxs->at(dev_id);
+    auto stream = nccl_ctx.stream();
+    auto comm = nccl_ctx.comm_;
+
+    VLOG(10) << "before all reduce run_order:" << run_order_
+             << "buffer:" << sendbuff << ", numel:" << count
+             << ", dev_id:" << dev_id << ", dtype:" << datatype
+             << ", place:" << place << ", stream:" << stream;
+
+    if (!FLAGS_sync_nccl_allreduce) {
+      cudaStreamWaitEvent(stream, inter_events_.at(dev_id), 0);
+    }
```

Comment:
`FLAGS_sync_nccl_allreduce` should not be used like this.

Reply:
Done.